File: core/capabilities/workflow/engine.py
from __future__ import annotations

import json
import uuid
import yaml
from pathlib import Path

from capabilities.workflow.models import FlowTemplate, FlowInstance


class WorkflowEngine:
    def __init__(self, workspace_path: str) -> None:
        self.workspace_path = workspace_path
        self.instance_dir = Path(workspace_path) / ".rhythm" / "flow_instances"
        self.instance_dir.mkdir(parents=True, exist_ok=True)
        
    def _template_dir(self) -> Path:
        d = Path(self.workspace_path) / "workflows"
        d.mkdir(parents=True, exist_ok=True)
        return d
        
    def save_template(self, template: FlowTemplate) -> None:
        target = self._template_dir() / f"{template.id}.yaml"
        with open(target, "w", encoding="utf-8") as f:
            yaml.safe_dump(template.to_dict(), f, allow_unicode=True, sort_keys=False, default_flow_style=False)
        
    def list_templates(self) -> list[FlowTemplate]:
        d = self._template_dir()
        res = []
        for p in d.glob("*.yaml"):
            try:
                with open(p, "r", encoding="utf-8") as f:
                    data = yaml.safe_load(f)
                res.append(FlowTemplate.from_dict(data))
            except Exception:
                pass
        for p in d.glob("*.json"):
            try:
                data = json.loads(p.read_text(encoding="utf-8"))
                res.append(FlowTemplate.from_dict(data))
            except Exception:
                pass
        return res
        
    def get_template(self, template_id: str) -> FlowTemplate | None:
        p_yaml = self._template_dir() / f"{template_id}.yaml"
        if p_yaml.exists():
            try:
                with open(p_yaml, "r", encoding="utf-8") as f:
                    data = yaml.safe_load(f)
                return FlowTemplate.from_dict(data)
            except Exception:
                pass
                
        p_json = self._template_dir() / f"{template_id}.json"
        if p_json.exists():
            try:
                return FlowTemplate.from_dict(json.loads(p_json.read_text(encoding="utf-8")))
            except Exception:
                pass
        return None

    def _instance_path(self, instance_id: str) -> Path:
        return self.instance_dir / f"{instance_id}.json"

    def get_instance(self, instance_id: str) -> FlowInstance | None:
        p = self._instance_path(instance_id)
        if p.exists():
            try:
                return FlowInstance.from_dict(json.loads(p.read_text(encoding="utf-8")))
            except Exception:
                return None
        return None
        
    def get_instances_for_session(self, session_id: str) -> list[FlowInstance]:
        res = []
        for p in self.instance_dir.glob("*.json"):
            try:
                inst = FlowInstance.from_dict(json.loads(p.read_text(encoding="utf-8")))
                if inst.session_id == session_id:
                    res.append(inst)
            except Exception:
                pass
        return res
        
    def save_instance(self, instance: FlowInstance) -> None:
        p = self._instance_path(instance.id)
        p.write_text(json.dumps(instance.to_dict(), ensure_ascii=False, indent=2), encoding="utf-8")
        
    def create_instance(self, session_id: str, template_id: str) -> FlowInstance:
        inst_id = str(uuid.uuid4())
        inst = FlowInstance(
            id=inst_id, 
            session_id=session_id, 
            template_id=template_id, 
            workspace_path=self.workspace_path,
            state="RUNNING"
        )
        self.save_instance(inst)
        return inst

File: src-tauri/src/core/capabilities/workflow/engine.rs
use std::path::PathBuf;
use std::fs;
use anyhow::Result;
use uuid::Uuid;
use crate::core::models::{FlowTemplate, FlowInstance};

pub struct WorkflowEngine {
    workspace_path: String,
    instance_dir: PathBuf,
}

impl WorkflowEngine {
    pub fn new(workspace_path: &str) -> Self {
        let mut path = PathBuf::from(workspace_path);
        path.push(".rhythm");
        path.push("flow_instances");
        fs::create_dir_all(&path).ok();
        Self {
            workspace_path: workspace_path.to_string(),
            instance_dir: path,
        }
    }

    fn template_dir(&self) -> PathBuf {
        let mut path = PathBuf::from(&self.workspace_path);
        path.push("workflows");
        fs::create_dir_all(&path).ok();
        path
    }

    pub fn save_template(&self, template: &FlowTemplate) -> Result<()> {
        let path = self.template_dir().join(format!("{}.yaml", template.id));
        let data = serde_yaml::to_string(template)?;
        fs::write(path, data)?;
        Ok(())
    }

    pub fn list_templates(&self) -> Result<Vec<FlowTemplate>> {
        let mut templates = vec![];
        if let Ok(entries) = fs::read_dir(self.template_dir()) {
            for entry in entries.flatten() {
                let path = entry.path();
                if let Some(ext) = path.extension() {
                    if ext == "yaml" || ext == "yml" {
                        if let Ok(data) = fs::read_to_string(&path) {
                            if let Ok(tpl) = serde_yaml::from_str(&data) {
                                templates.push(tpl);
                            }
                        }
                    } else if ext == "json" {
                        if let Ok(data) = fs::read_to_string(&path) {
                            if let Ok(tpl) = serde_json::from_str(&data) {
                                templates.push(tpl);
                            }
                        }
                    }
                }
            }
        }
        Ok(templates)
    }

    pub fn get_template(&self, template_id: &str) -> Result<Option<FlowTemplate>> {
        let yaml_path = self.template_dir().join(format!("{}.yaml", template_id));
        if yaml_path.exists() {
            let data = fs::read_to_string(yaml_path)?;
            return Ok(Some(serde_yaml::from_str(&data)?));
        }
        let json_path = self.template_dir().join(format!("{}.json", template_id));
        if json_path.exists() {
            let data = fs::read_to_string(json_path)?;
            return Ok(Some(serde_json::from_str(&data)?));
        }
        Ok(None)
    }

    pub fn save_instance(&self, instance: &FlowInstance) -> Result<()> {
        let path = self.instance_dir.join(format!("{}.json", instance.id));
        let data = serde_json::to_string_pretty(instance)?;
        fs::write(path, data)?;
        Ok(())
    }

    pub fn get_instance(&self, instance_id: &str) -> Result<Option<FlowInstance>> {
        let path = self.instance_dir.join(format!("{}.json", instance_id));
        if path.exists() {
            let data = fs::read_to_string(path)?;
            return Ok(Some(serde_json::from_str(&data)?));
        }
        Ok(None)
    }

    pub fn get_instances_for_session(&self, session_id: &str) -> Result<Vec<FlowInstance>> {
        let mut instances = vec![];
        if let Ok(entries) = fs::read_dir(&self.instance_dir) {
            for entry in entries.flatten() {
                if let Ok(data) = fs::read_to_string(entry.path()) {
                    if let Ok(inst) = serde_json::from_str::<FlowInstance>(&data) {
                        if inst.session_id == session_id {
                            instances.push(inst);
                        }
                    }
                }
            }
        }
        Ok(instances)
    }

    pub fn create_instance(&self, session_id: &str, template_id: &str) -> Result<FlowInstance> {
        let inst = FlowInstance {
            id: Uuid::new_v4().to_string(),
            session_id: session_id.to_string(),
            template_id: template_id.to_string(),
            workspace_path: self.workspace_path.clone(),
            current_step_index: 0,
            state: "RUNNING".to_string(),
            context_data: std::collections::HashMap::new(),
            pending_question: None,
            required_keys: vec![],
        };
        self.save_instance(&inst)?;
        Ok(inst)
    }
}
